muxtools/subtitle/font.py

```diff
@@ -1,14 +1,274 @@
+import os
 import re
 import shutil
 import logging
 from pathlib import Path
 from font_collector import ABCFontFace, VariableFontFace
+from fontTools.subset import Subsetter
+from fontTools import ttLib
+import time
+import hashlib
+import base64
 
 from .sub import SubFile, FontFile as MTFontFile
+from ..utils.convert import sizeof_fmt
 from ..utils.env import get_workdir
 from ..utils.log import warn, error, info, danger
 
 
+__all__ = [
+    "subset_fonts",
+]
+
+# A selection of common unicode characters to always include when subsetting fonts
+# Follows the format: 'U+XXXX'
+# For ranges, use 'U+XXXX-YYYY'
+UNFORMATTED_COMMON_UNICODE_CHARS = [
+    'U+0000-00FF',  # Basic Latin + Latin-1 Supplement
+    'U+0100-024F',  # Latin Extended-A + Latin Extended-B
+    'U+1E00-1EFF',  # Latin Extended Additional
+    'U+2000-206F',  # General Punctuation
+    'U+20A0-20CF',  # Currency Symbols
+]
+
+def _parse_unicode_chars(char_list: list[str]) -> list[str]:
+    """Parse unicode character definitions including ranges."""
+    result = []
+    for item in char_list:
+        if item.startswith('U+'):
+            if '-' in item:
+                # Handle range
+                parts = item.replace('U+', '').split('-')
+                start = int(parts[0], 16)
+                end = int(parts[1], 16)
+                result.extend([chr(i) for i in range(start, end + 1)])
+            else:
+                # Single character
+                result.append(chr(int(item.replace('U+', ''), 16)))
+        else:
+            if len(item) == 1:
+                result.append(item)
+            else:
+                warn(f"Invalid unicode character format: {item}", _parse_unicode_chars)
+
+    return result
+
+COMMON_UNICODE_CHARS = _parse_unicode_chars(UNFORMATTED_COMMON_UNICODE_CHARS)
+
+
+def _hash_font_name(font_name: str, run_time: str) -> str:
+    font_name = font_name.replace(" ", "").strip()
+
+    hash = base64.urlsafe_b64encode(
+        hashlib.sha256(f"{font_name}_Subset_{run_time}".encode('utf-8')).digest()
```

Comment:
IIUC this ensures a unique font name every time this runs. It’s good that it won’t clash with other fonts, but it may be better to use the subsetting parameters (i. e. the set of codepoints, and anything else if you have any other parameters) instead of `run_time` so that identically-subset fonts end up identical and can be deduplicated.